core/src/subgraph/registrar.rs

```diff
@@ -0,0 +1,776 @@
+                    previous_version_hash.clone().into(),
+                )))?;
+            let referencing_version_entity_ids = referencing_version_entities
+                .iter()
+                .map(|entity| entity.id().unwrap())
+                .collect::<Vec<_>>();
+
+            // Find all subgraphs that have one of these versions as currentVersion
+            let subgraphs_with_current_version_referencing_hash = store.find(
+                SubgraphEntity::query().filter(EntityFilter::In(
+                    "currentVersion".to_owned(),
+                    referencing_version_entity_ids
+                        .into_iter()
+                        .map(Value::from)
+                        .collect(),
+                )),
+            )?;
+            let subgraph_ids_with_current_version_referencing_hash =
+                subgraphs_with_current_version_referencing_hash
+                    .iter()
+                    .map(|entity| entity.id().unwrap())
+                    .collect::<Vec<_>>();
+
+            // If this subgraph is the only one with this hash as the current version
+            if subgraph_ids_with_current_version_referencing_hash.len() == 1 {
+                ops.push(EntityOperation::Remove {
+                    key: SubgraphDeploymentEntity::key(
+                        SubgraphId::new(previous_version_hash).unwrap(),
+                    ),
+                });
+            }
+        }
+    }
+
+    // Check if deployment already exists for this hash
+    let deployment_entity_opt = store.get(SubgraphDeploymentEntity::key(manifest.id.clone()))?;
+    ops.push(EntityOperation::AbortUnless {
+        description: "Subgraph deployment entity must continue to exist/not exist".to_owned(),
+        query: SubgraphDeploymentEntity::query().filter(EntityFilter::Equal(
+            "id".to_owned(),
+            manifest.id.to_string().into(),
+        )),
+        entity_ids: if deployment_entity_opt.is_some() {
+            vec![manifest.id.to_string()]
+        } else {
+            vec![]
+        },
+    });
+
+    // Create deployment entity only if it does not exist already
+    if deployment_entity_opt.is_none() {
+        ops.extend(SubgraphDeploymentEntity::new(node_id, false).write_operations(&manifest.id));
+    }
+
+    // TODO support delayed update of currentVersion
+    ops.extend(SubgraphEntity::update_current_version_operations(
+        &subgraph_entity_id,
+        &version_entity_id,
+    ));
+
+    // Commit entity ops
+    store.apply_entity_operations(ops, EventSource::None)?;
+
+    Ok(())
+}
+
+fn remove_subgraph(
+    logger: &Logger,
+    store: Arc<impl Store>,
+    name: SubgraphName,
+) -> Result<(), SubgraphRegistrarError> {
+    debug!(logger, "Removing subgraph: {:?}", name.to_string());
+
+    let mut ops = vec![];
+
+    // Find the subgraph entity
+    let matching_subgraph_entities = store
+        .find(SubgraphEntity::query().filter(EntityFilter::Equal(
```

Comment:
Makes sense. I just pushed a commit to add `store.find_one`, which sets a range and returns `Option<Entity>` instead of `Vec<Entity>`, and refactored these `find` calls to use `find_one` instead